### PR TITLE
Implement basic DecentralizedOptimizer

### PR DESCRIPTION
**Pull request by justheuristic**
This PR implements a simple DecentralizedOptimizer that averages model parameters with peers. It is meant as a base implementation that will will be extended with #184 and a collaborative-training-like optimizer in subsequent PRs.

### CODE REVIEW

File: hivemind/client/averaging/__init__.py [hunk 1]
```diff
@@ -196,14 +197,14 @@ def run_in_background(self, await_ready=True, timeout=None):
...
-        if self.is_alive():
+        if self._parent_pid != os.getpid() or self.is_alive():
             self._pipe.send(('_SHUTDOWN', None))
             self.terminate()
         else:
             logger.warning("DHT shutdown has no effect: the process is not alive")
 
     def __del__(self):
-        if self.is_alive():
+        if self._parent_pid != os.getpid() or self.is_alive():
```
Comment (justheuristic) [inline hivemind/client/averaging/__init__.py:207]: This check fixes an edge case where deleting muiltiple averagers at once may trigger AssertionError "Can only check is_alive for child processes"

**Comment by justheuristic** [comment]
TODOS:
* add a more advanced test (aka benchmark) in a PR with full optimizer
* update sphinx docs
* [refactoring] maybe push load_state_from_peers from averager into DecentralizedOptimizer

File: tests/test_training.py [hunk 2]
```diff
@@ -36,3 +37,31 @@ def test_training(max_steps: int = 100, threshold: float = 0.9):
                 break
 
         assert accuracy >= threshold, f"too small accuracy: {accuracy}"
+
+
+@pytest.mark.forked
+def test_decentralized_optimizer_step():
```
Comment (yhn112) [inline tests/test_training.py:43]: I would like to have a more complex test here, but sure, we don't need it immediately.

File: hivemind/client/optim/simple.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+class DecentralizedOptimizer(DecentralizedOptimizerBase):
+    """
+    A simple optimizer that trains a shared model by averaging model parameters with peers in the background.
+
+    :param opt: a pytorch optimizer configured to update model parameters.
+    :param dht: a running hivemind DHT daemon connected to other peers
+    :param prefix: all DHT keys that point to optimization metadata will have this prefix
+    :param target_group_size: maximum group size for averaging (see DecentralizedAverager)
+    :param kwargs: additional parameters passed to DecentralizedAverager
```
Comment (mryab) [inline hivemind/client/optim/simple.py:22]: We need an explicit argument for update frequency in the __init__ for this class, so that it isn't necessary to look for the frequency in other docstrings

File: tests/test_training.py [hunk 4]
```diff
@@ -36,3 +37,31 @@ def test_training(max_steps: int = 100, threshold: float = 0.9):
...
+
+    param1 = torch.nn.Parameter(torch.zeros(32, 32), requires_grad=True)
+    opt1 = DecentralizedOptimizer(opt=torch.optim.SGD([param1], lr=0.1),
+                                  dht=DHT(initial_peers=initial_peers, start=True),
+                                  prefix='foo', target_group_size=2, verbose=True)
+
+    param2 = torch.nn.Parameter(torch.ones(32, 32), requires_grad=True)
+    opt2 = DecentralizedOptimizer(opt=torch.optim.SGD([param2], lr=0.05),
+                                  dht=DHT(initial_peers=initial_peers, start=True),
+                                  prefix='foo', target_group_size=2, verbose=True)
```
Comment (mryab) [inline tests/test_training.py:55]: Also, pass the update frequency to these optimizers to ensure consistency with `time.sleep` below

File: hivemind/client/optim/simple.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            if group_info is not None:
+                with torch.no_grad(), averager.get_tensors() as averaged_tensors:
+                    for local_tensor, averaged_tensor in zip(local_tensors, averaged_tensors):
+                        local_tensor[...] = averaged_tensor.to(dtype=local_tensor.dtype)
+                if verbose:
+                    logger.info(f"Finished averaging round in with {len(group_info)} peers.")
+            else:
+                logger.info(f"Averaging round failed: could not find group.")
+        except Exception as e:
+            logger.info(f"Averaging round failed: caught {e}.")
```
Comment (mryab) [inline hivemind/client/optim/simple.py:88]: `logger.error`?

File: hivemind/client/optim/simple.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            group_info = averager.step(**kwargs)
+
+            if group_info is not None:
+                with torch.no_grad(), averager.get_tensors() as averaged_tensors:
+                    for local_tensor, averaged_tensor in zip(local_tensors, averaged_tensors):
+                        local_tensor[...] = averaged_tensor.to(dtype=local_tensor.dtype)
+                if verbose:
+                    logger.info(f"Finished averaging round in with {len(group_info)} peers.")
+            else:
+                logger.info(f"Averaging round failed: could not find group.")
```
Comment (mryab) [inline hivemind/client/optim/simple.py:86]: `logger.warning`?

File: hivemind/client/optim/simple.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+        try:
+            if verbose:
+                logger.info(f"Starting a new averaging round with current parameters.")
+            group_info = averager.step(**kwargs)
+
+            if group_info is not None:
+                with torch.no_grad(), averager.get_tensors() as averaged_tensors:
+                    for local_tensor, averaged_tensor in zip(local_tensors, averaged_tensors):
+                        local_tensor[...] = averaged_tensor.to(dtype=local_tensor.dtype)
```
Comment (mryab) [inline hivemind/client/optim/simple.py:82]: It's best to explicitly write `copy_()` for the local tensor

File: hivemind/client/optim/simple.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        update_event.wait()
+        update_event.clear()
+        if stop_event.is_set():
+            break
+
+        with torch.no_grad(), averager.get_tensors() as averaged_tensors:
+            local_tensors = tuple(p for group in opt.param_groups for p in group['params'])
+            assert len(local_tensors) == len(averaged_tensors), "The number of optimized parameters should not change."
+            for local_tensor, averaged_tensor in zip(local_tensors, averaged_tensors):
+                averaged_tensor[...] = local_tensor.cpu().float()
```
Comment (mryab) [inline hivemind/client/optim/simple.py:72]: 1) It's best to explicitly write copy_() for the local tensor
2) What if the dtype of averaged_tensor isn't float?

File: hivemind/client/optim/simple.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    def __del__(self):
+        self.stop_event.set()
+        while self.background_averaging_thread.is_alive():
+            self.update_event.set()
+
+
+def _average_parameters_in_background(update_event: Event, stop_event: Event, averager: DecentralizedAverager,
+                                      opt: torch.optim.Optimizer, verbose: bool, **kwargs):
+    """ Iteratively find groups of peers, average parameters with these peers and update local model parameters. """
+    while True:
```
Comment (mryab) [inline hivemind/client/optim/simple.py:62]:
```suggestion
    while not stop_event.is_set():
```

Comment (mryab) [reply hivemind/client/optim/simple.py:62]: This way, you can get rid of `break` below

File: hivemind/client/optim/simple.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        update_event.wait()
+        update_event.clear()
+        if stop_event.is_set():
+            break
+
+        with torch.no_grad(), averager.get_tensors() as averaged_tensors:
```
Comment (mryab) [inline hivemind/client/optim/simple.py:68]: I think this entire function can be wrapped in torch.no_grad

File: hivemind/client/optim/simple.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                                                  kwargs=dict(verbose=verbose, timeout=timeout))
+        self.background_averaging_thread.start()
+
+    def step(self, closure: Optional[Callable[[], Any]] = None):
+        self.update_event.set()
+        with self.lock_parameters:
+            return self.opt.step(closure)
+
+    def zero_grad(self, set_to_none: Optional[bool] = False):
+        return self.opt.zero_grad()
```
Comment (mryab) [inline hivemind/client/optim/simple.py:51]: `set_to_none` isn't passed to the original optimizer

File: hivemind/client/optim/simple.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    :param prefix: all DHT keys that point to optimization metadata will have this prefix
+    :param target_group_size: maximum group size for averaging (see DecentralizedAverager)
+    :param kwargs: additional parameters passed to DecentralizedAverager
+
+    :note: using DecentralizedOptimizer with adaptive learning rates may result in poor convergence due to
+      out-of-sync adaptive learning rates (such as adam second momentum or schedule step). Please ensure that these
+      statistics are synchronized or use a more advanced DecentralizedOptimizer version, if applicable.
+    :note: the base optimizer cannot add param groups after the DecentralizedOptimizer is created
+    """
+    averager_step: Optional[MPFuture] = None
```
Comment (mryab) [inline hivemind/client/optim/simple.py:29]: Unneeded variable?

File: hivemind/client/optim/simple.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+from hivemind.dht import DHT
+from hivemind.client.averaging import DecentralizedAverager
+from hivemind.client.optim.base import DecentralizedOptimizerBase
+from hivemind.utils import MPFuture, get_logger
+
+logger = get_logger(__name__)
+
+
+class DecentralizedOptimizer(DecentralizedOptimizerBase):
```
Comment (mryab) [inline hivemind/client/optim/simple.py:14]: `ParameterAveragingOptimizer`? (just to clarify that this optimizer only does parameter averaging)

Comment (mryab) [reply hivemind/client/optim/simple.py:14]: Alternatively, we can have one DecentralizedOptimizer class which averages parameters,gradients or statistics depending on use case. It shouldn't be too large, given that most of the work is done in DecentralizedAverager

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/client/optim/simple.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            return self.opt.step(closure)
+
+    def zero_grad(self, set_to_none: Optional[bool] = False):
+        return self.opt.zero_grad()
+
+    def __del__(self):
+        self.stop_event.set()
+        while self.background_averaging_thread.is_alive():
+            self.update_event.set()
+
```
Comment (nevec) [inline hivemind/client/optim/simple.py:57]: I'd like to have an explicit method for shutting down the averager

**Review by nevec — APPROVED** [review]

File: tests/test_training.py (hunk 2 above)
Comment (justheuristic) [reply tests/test_training.py:43]: Resolution: will add a more complex test for a more advanced optimizer in a subsequent PR.

File: hivemind/client/optim/base.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+    def add_param_group(self, param_group: dict) -> None:
+        raise ValueError(f"{self.__class__.__name__} does not support calling add_param_group after creation."
+                         f"Please provide all parameter groups at init.")
+
+    def state_dict(self) -> dict:
+        return self.opt.state_dict()
+
+    def load_state_dict(self, state_dict: dict) -> None:
+        return self.opt.load_state_dict(state_dict)
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/optim/base.py:27]: it was my mistake that load_state_dict returned None. In actuality, some optimizers report statuses like "all keys matched"

File: hivemind/client/optim/simple.py (hunk 3 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: Solution: implemented averaging_period, measured in seconds. Here's how it works:

If the optimizer is active, the background thread will await until get_dht_time is divisible by averaging_period, then begin averaging.

This trick ensures that all peers begin averaging at almost the same time. Otherwise, having a long averaging_period will result in a situation where only a small fraction of peers is active at any given point in time and no peer is able to find group (to say nothing about following the moshpit protocol)

File: tests/test_training.py (hunk 4 above)
Comment (justheuristic) [reply tests/test_training.py:55]: Alas, the time.sleep is meant not for consistency with averaging period, but simply to make sure that the slow CI executor was able to... well, do the averaging :)

File: hivemind/client/optim/simple.py (hunk 5 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: Fixed

File: hivemind/client/optim/simple.py (hunk 9 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py:68]: Alas, it aint that simple:
* It is important that `__del__` handler will terminate the thread immediately without running an extra round of averaging
* We could move update_event.wait/clear to the end of the cycle, but we must wait for the first update to run the averaging

Which one do you prefer?
- move update_event.wait()/clear() to the end of the cycle, add an extra wait()/clear() before the cycle
- check for stop_event twice - in the `while` condition _and_ directly after wait()/clear()
- keep the code as is?

File: hivemind/client/optim/simple.py (hunk 11 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py:54]: nice catch!

File: hivemind/client/optim/simple.py (hunk 13 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: Renamed (first option)

While most of the work is indeed handled by the DecentralizedAverager, i am unsure which user inter would be preferable for the generalized optimizer (with local gradients, statistics, etc).

Let us handle that in subsequent pull requests.

File: hivemind/client/optim/simple.py (hunk 12 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: thanks!

File: hivemind/client/optim/simple.py (hunk 6 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: agreed

File: hivemind/client/optim/simple.py (hunk 10 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: done

File: hivemind/client/optim/simple.py (hunk 13 above)
Comment (justheuristic) [reply hivemind/client/optim/simple.py]: Also: added a DecentralizedSGD ([term usage example](https://arxiv.org/pdf/2003.10422.pdf)) wrapper for convenience -- mainly not to scare people off with ParameterAveragingOptimizer. To the best of my knowledge, SGD is the only mainstream optimizer from torch.optim that is guaranteed to converge with parameter averaging.

Note: i intend to use DecentralizedSGD for resnet50 example after some performance tweaks.

**Review by mryab — APPROVED** [review]
